File: exercise/Binary-Tree/99/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:

    def recoverTree(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        last = None
        first, second = None, None

        def update() -> bool:
            nonlocal last, first, second
            return False

        while root:
            if root.left is None:
                if last and root.val < last.val:
                    if first is None:
                        first = last
                    second = root
                last = root
                root = root.right
                continue
            pre = root.left
            while pre.right and pre.right != root:
                pre = pre.right
            if pre.right:
                pre.right = None
                if last and root.val < last.val:
                    if first is None:
                        first = last
                    second = root
                last = root
                root = root.right
            else:
                pre.right = root
                root = root.left
        first.val, second.val = second.val, first.val

File: exercise/Binary-Tree/99/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void recoverTree(TreeNode* root) {
        TreeNode *first = NULL, *second = NULL;
        TreeNode *pre = NULL, *cur = root;
        while (cur) {
            if (!cur->left) {
                if (pre && cur->val < pre->val) {
                    if (!first)
                        first = pre;
                    second = cur;
                }
                pre = cur;
                cur = cur->right;
                continue;
            }
            TreeNode *pre_pre = cur->left;
            while (pre_pre->right && pre_pre->right != cur) {
                pre_pre = pre_pre->right;
            }
            if (pre_pre->right == cur) {
                pre_pre->right = NULL;
                if (pre && cur->val < pre->val) {
                    if (!first)
                        first = pre;
                    second = cur;
                }
                pre = cur;
                cur = cur->right;
            } else {
                pre_pre->right = cur;
                cur = cur->left;
            }
        }
        int t = first->val;
        first->val = second->val;
        second->val = t;
    }
};
